core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/principal/Principal.scala

```diff
@@ -264,79 +266,94 @@ class Principal(val metadata: OpExecConfig) extends Actor with ActorLogging with
+                  1.milliseconds,
+                  30.seconds,
+                  self,
+                  EnforceStateCheck
+                )
+                context.parent ! ReportState(PrincipalState.CollectingBreakpoints)
+                context.become(collectingBreakpoints)
+                unstashAll()
               }
-              context.system.scheduler
-                .scheduleOnce(tau, () => unCompletedWorkers.foreach(worker => worker ! Pause))
-              safeRemoveAskHandle()
-              periodicallyAskHandle =
-                context.system.scheduler.schedule(30.seconds, 30.seconds, self, EnforceStateCheck)
-              context.become(pausing)
-              unstashAll()
-            }
-          case WorkerState.Paused =>
-            if (whenAllWorkersCompleted) {
-              safeRemoveAskHandle()
-              context.parent ! ReportState(PrincipalState.Completed)
-              context.become(completed)
-              unstashAll()
-            } else if (whenAllUncompletedWorkersBecome(WorkerState.Paused)) {
-              safeRemoveAskHandle()
-              context.parent ! ReportState(PrincipalState.Paused)
-              context.become(paused)
-              unstashAll()
-            } else if (
-              unCompletedWorkerStates
-                .forall(x => x == WorkerState.Paused || x == WorkerState.LocalBreakpointTriggered)
-            ) {
-              workersTriggeredBreakpoint =
-                workerStateMap.filter(_._2 == WorkerState.LocalBreakpointTriggered).keys
-              safeRemoveAskHandle()
-              periodicallyAskHandle = context.system.scheduler.schedule(
-                1.milliseconds,
-                30.seconds,
-                self,
-                EnforceStateCheck
-              )
-              context.parent ! ReportState(PrincipalState.CollectingBreakpoints)
-              context.become(collectingBreakpoints)
-              unstashAll()
-            }
-          case WorkerState.Completed =>
-            if (whenAllWorkersCompleted) {
-              if (timer.isRunning) {
-                timer.stop()
+            case WorkerState.Completed =>
+              if (whenAllWorkersCompleted) {
+                if (timer.isRunning) {
+                  timer.stop()
+                }
+                log.info(metadata.tag.toString + " completed! Time Elapsed: " + timer.toString())
+                context.parent ! ReportState(PrincipalState.Completed)
+                context.become(completed)
+                unstashAll()
               }
-              log.info(metadata.tag.toString + " completed! Time Elapsed: " + timer.toString())
-              context.parent ! ReportState(PrincipalState.Completed)
-              context.become(completed)
-              unstashAll()
-            }
-          case _ => //skip others for now
+            case _ => //skip others for now
+          }
         }
+      } catch {
+        case e: WorkflowRuntimeException =>
```

Comment:
I believe adding this try catch changes the actor behavior upon throwing an exception.
We add a try-catch here -> then send message to controller -> then controller just invoke the event listener and then continue. Now the system is just in a bad state and anything could happen, it could just lead to a series of more errors, or they will hang there, and we just don't know what the actors will even do. 

I believe in most of the cases, upon such exceptions, we just want to kill the whole workflow. I think this goes hand-in-hand with the failure model in Akka. Can we take a look at how Akka does it and how can we leverage Akka's existing fault handling system?

Related akka documentation: https://doc.akka.io/docs/akka/current/fault-tolerance.html